.github/workflows/main.yml

```diff
@@ -0,0 +1,74 @@
+name: CI
+
+# Controls when the action will run. Triggers the workflow on push or pull request
+# events but only for the master branch
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  build-toolchain:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-16.04, ubuntu-18.04, ubuntu-20.04]
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        # we don't need submodules for building the toolchain
+
+      # toolchain takes ~40min to build, so we want to avoid that whenever possible!
+      - uses: actions/cache@v2
+        id: toolchain-cache
+        with:
+          path: toolchain.tar.gz
+          # gcc/binutils/etc versions are in build_toolchain.sh, so that'll be part of our cache key
+          key: ${{ matrix.os }}-${{ hashFiles('ndless-sdk/toolchain/build_toolchain.sh') }}
+      - name: Install dependencies
+        if: steps.toolchain-cache.outputs.cache-hit != 'true'
+        run: sudo apt install -y libmpfr-dev libmpc-dev
+      - name: Build
+        if: steps.toolchain-cache.outputs.cache-hit != 'true'
+        working-directory: ndless-sdk/toolchain
+        run: ./build_toolchain.sh
+      - name: Package toolchain
+        if: steps.toolchain-cache.outputs.cache-hit != 'true'
+        run: tar -cvzf toolchain.tar.gz ndless-sdk/toolchain/install
+      
+      - name: Upload built toolchain
```

Comment:
Ideally (if accessing previous artifacts were possible) we could skip uploading the toolchain if it wasn't rebuilt.

In the end I agree (doing it in one go is what we do with github.com/shinyblink/sled anyway) that it's the better solution with regards to runtime, simply because of the artifact overhead.

I'll merge the two jobs, and remove the download part. While uploading the toolchain might be beneficial for developers on CI supported OSs, downloading it again in the same job is unnessecary: the cache mechanism takes care of that much faster.